classes/SubjectDemographics.php

```diff
@@ -89,54 +88,51 @@ public function setAllSubjects(): void
-                if ($response->getStatusCode() < 300) {
-                    $data = json_decode($response->getBody(), true);
-                    if (empty($data)) {
-                        return [];
-                    } else {
-                        // create entity table record before return.
-                        #$entity = (new Entities)->getFactory()->create(OnCoreIntegration::ONCORE_SUBJECTS, $data);
-                        if($demo){
-                            $entity = (new Entities)->update(OnCoreIntegration::ONCORE_SUBJECTS, $demo['id'], $data);
-                        }
-                        else{
-                            $entity = (new Entities)->create(OnCoreIntegration::ONCORE_SUBJECTS, $data);
-                        }
-                        if ($entity) {
-                            return $data;
-                        } else {
-                            Entities::createLog('cant create subject record for ' . $data['subjectDemographicsId']);
-                            Entities::createLog(implode('<br>', $data));
-                        }
-
-                    }
+    public function getOnCoreSubjectDemographicsAsync($client, $subjectDemographicsId, $forceDemographicsPull = false)
+    {
+        return new Promise\Promise(function () use ($client, $subjectDemographicsId, $forceDemographicsPull, &$promise) {
+            try {
+                $demo = $this->getAllSubjects()[$subjectDemographicsId] ?? null;
+                if (empty($demo) || $forceDemographicsPull) {
+
+                    $client->getAsync('subjectDemographics/' . $subjectDemographicsId)
+                        ->then(function ($response) use (&$promise, $demo) {
+                            if ($response->getStatusCode() < 300) {
+                                $data = json_decode($response->getBody(), true);
+                                if (empty($data)) {
+                                    $promise->resolve([]);
+                                } else {
+                                    if ($demo) {
+                                        (new Entities)->update(OnCoreIntegration::ONCORE_SUBJECTS, $demo['id'], $data);
+                                    } else {
+                                        (new Entities)->create(OnCoreIntegration::ONCORE_SUBJECTS, $data);
+                                    }
+                                    $promise->resolve($data);
+                                }
+                            }
+                        })
+                        ->otherwise(function ($e) use (&$promise) {
+                            $promise->reject($e);
+                        });
+                } else {
+                    $promise->resolve($demo);
                 }
-            } else {
-                return $demo;
+            } catch (GuzzleException $e) {
+                $promise->reject(new \Exception(json_decode($e->getResponse()->getBody()->getContents(), true)['message']));
```

Comment:
Calling $e->getResponse()->getBody()->getContents() without checking if getResponse() is null may cause additional errors.

```
$message = $e->getResponse() ? json_decode($e->getResponse()->getBody()->getContents(), true)['message'] ?? 'Unknown error' : $e->getMessage();
            $promise->reject(new \Exception($message));

```